internal/controllers/machinedeployment/machinedeployment_status.go

```diff
@@ -0,0 +1,502 @@
+		return
+	}
+
+	v1beta2conditions.Set(machineDeployment, metav1.Condition{
+		Type:    remediatingCondition.Type,
+		Status:  metav1.ConditionTrue,
+		Reason:  clusterv1.MachineDeploymentRemediatingV1Beta2Reason,
+		Message: remediatingCondition.Message,
+	})
+}
+
+func setDeletingCondition(_ context.Context, machineDeployment *clusterv1.MachineDeployment, machineSets []*clusterv1.MachineSet, machines collections.Machines, getAndAdoptMachineSetsForDeploymentSucceeded bool) {
+	// If we got unexpected errors in listing the machines sets or machines (this should never happen), surface them.
+	if !getAndAdoptMachineSetsForDeploymentSucceeded || machines == nil {
+		v1beta2conditions.Set(machineDeployment, metav1.Condition{
+			Type:    clusterv1.MachineDeploymentDeletingV1Beta2Condition,
+			Status:  metav1.ConditionUnknown,
+			Reason:  clusterv1.MachineDeploymentDeletingInternalErrorV1Beta2Reason,
+			Message: "Please check controller logs for errors",
+		})
+		return
+	}
+
+	if machineDeployment.DeletionTimestamp.IsZero() {
+		v1beta2conditions.Set(machineDeployment, metav1.Condition{
+			Type:   clusterv1.MachineDeploymentDeletingV1Beta2Condition,
+			Status: metav1.ConditionFalse,
+			Reason: clusterv1.MachineDeploymentDeletingDeletionTimestampNotSetV1Beta2Reason,
+		})
+		return
+	}
+
+	message := ""
+	if len(machines) > 0 {
+		message = fmt.Sprintf("Deleting %d replicas", len(machines))
+		staleMessage := aggregateStaleMachines(machines)
+		if staleMessage != "" {
+			message += fmt.Sprintf(" and %s", staleMessage)
+		}
+	}
+	if len(machines) == 0 && len(machineSets) > 0 {
+		// Note: this should not happen or happen for a very short time while the finalizer is removed.
+		message = fmt.Sprintf("Deleting %d MachineSets", len(machineSets))
+	}
+	v1beta2conditions.Set(machineDeployment, metav1.Condition{
+		Type:    clusterv1.MachineDeploymentDeletingV1Beta2Condition,
+		Status:  metav1.ConditionFalse,
+		Reason:  clusterv1.MachineDeploymentDeletingDeletionTimestampSetV1Beta2Reason,
+		Message: message,
+	})
+}
+
+func calculateMissingReferencesMessage(machineDeployment *clusterv1.MachineDeployment, bootstrapTemplateNotFound, infraMachineTemplateNotFound bool) string {
+	missingObjects := []string{}
+	if bootstrapTemplateNotFound {
+		missingObjects = append(missingObjects, machineDeployment.Spec.Template.Spec.Bootstrap.ConfigRef.Kind)
+	}
+	if infraMachineTemplateNotFound {
+		missingObjects = append(missingObjects, machineDeployment.Spec.Template.Spec.InfrastructureRef.Kind)
+	}
+
+	if len(missingObjects) == 0 {
+		return ""
+	}
+
+	if len(missingObjects) == 1 {
+		return fmt.Sprintf("because %s does not exist", missingObjects[0])
+	}
+
+	return fmt.Sprintf("because %s do not exist", strings.Join(missingObjects, " and "))
+}
+
+func aggregateStaleMachines(machines collections.Machines) string {
+	if len(machines) == 0 {
+		return ""
+	}
```

Comment:
I know we're probably going to de-duplicate soon. But maybe let's add this to the other 2 copies of the func for now already